IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m

```diff
@@ -26,21 +26,375 @@
 //------------------------------------------------------------------------------
 
 #import "MSIDOAuth2EmbeddedWebviewController.h"
+#import "MSIDTelemetry+Internal.h"
+#import "MSIDTelemetryUIEvent.h"
+#import "MSIDTelemetryEventStrings.h"
+#import "MSIDAadAuthorityCache.h"
+#import "MSIDError.h"
+#import "MSIDWebOAuth2Response.h"
+#import "MSIDWebviewAuthorization.h"
+#import "MSIDChallengeHandler.h"
+#import "MSIDNTLMHandler.h"
+#import "MSIDAuthority.h"
+#import "MSIDWorkPlaceJoinConstants.h"
+
+#if TARGET_OS_IPHONE
+#import "MSIDAppExtensionUtil.h"
+#import "MSIDPKeyAuthHandler.h"
+#else
+#define DEFAULT_WINDOW_WIDTH 420
+#define DEFAULT_WINDOW_HEIGHT 650
+#endif
 
 @implementation MSIDOAuth2EmbeddedWebviewController
+{
+    NSURL *_startUrl;
+    NSURL *_endUrl;
+    void (^_completionHandler)(MSIDWebOAuth2Response *response, NSError *error);
+    
+    NSLock *_completionLock;
+    NSTimer *_spinnerTimer; // Used for managing the activity spinner
+}
 
+- (id)initWithStartUrl:(NSURL *)startUrl
+                endURL:(NSURL *)endUrl
+               webview:(WKWebView *)webview
+               context:(id<MSIDRequestContext>)context
+{
+    self = [super initWithContext:context];
+    
+    if (self)
+    {
+        self.webView = webview;
+        _startUrl = startUrl;
+        _endUrl = endUrl;
+        
+        _completionLock = [[NSLock alloc] init];
+        self.complete = NO;
+    }
+    
+    return self;
+}
+
+-(void)dealloc
+{
+    [self.webView setNavigationDelegate:nil];
+    self.webView = nil;
+}
+
+- (BOOL)startWithCompletionHandler:(MSIDWebUICompletionHandler)completionHandler
+{
+    // If we're not on the main thread when trying to kick up the UI then
+    // dispatch over to the main thread.
+    if (![NSThread isMainThread])
+    {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self startWithCompletionHandler:completionHandler];
+        });
+        return YES;
+    }
+    
+    // Save the completion block
+    _completionHandler = [completionHandler copy];
+    
+    NSError *error = nil;
+    [self loadView:&error];
+    if (error)
+    {
+        [self endWebAuthenticationWithError:error orURL:nil];
+        return YES;
+    }
+    
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:_startUrl];
 #if TARGET_OS_IPHONE
-@synthesize  parentViewController;
+    // Currently Apple has a bug in iOS about WKWebview handling NSURLAuthenticationMethodClientCertificate.
+    // It swallows the challenge response rather than sending it to server.
+    // Therefore we work around the bug by using PKeyAuth for WPJ challenge in iOS
+    [request addValue:kMSIDPKeyAuthHeaderVersion forHTTPHeaderField:kMSIDPKeyAuthHeader];
```

Comment:
this should be appended only for AAD request. Consider making the logic someone else's responsibility, e.g. passing custom headers in addition to the start URL or maybe even the whole URL request.

Reply:
done